test/libyul/yulOptimizerTests/fullSuite/abi2.yul

```diff
@@ -0,0 +1,1218 @@
+				converted := cleanup_assert_t_enum$_Operation_$1949(value)
+			}
+		
+				function copy_calldata_to_memory(src, dst, length) {
+					calldatacopy(dst, src, length)
+					// clear end
+					mstore(add(dst, length), 0)
+				}
+			
+				function copy_memory_to_memory(src, dst, length) {
+					let i := 0
+					for { } lt(i, length) { i := add(i, 32) }
+					{
+						mstore(add(dst, i), mload(add(src, i)))
+					}
+					if gt(i, length)
+					{
+						// clear end
+						mstore(add(dst, length), 0)
+					}
+				}
+			
+			function round_up_to_mul_of_32(value) -> result {
+				result := and(add(value, 31), not(31))
+			}
+			}
+// ----
+// fullSuite
+// {
+//     {
+//         let _1 := 1
+//         let _2 := mload(_1)
+//         let _3 := 0
+//         let _1017 := mload(_3)
+//         let abi_decode_value0_156 := _3
+//         if slt(sub(_2, _1017), 64)
+//         {
+//             revert(_3, _3)
+//         }
+//         {
+//             abi_decode_value0_156 := and(calldataload(_1017), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+//         }
+//         sstore(_3, abi_decode_value0_156)
+//         let _1019 := mload(8)
+//         let _1021 := mload(7)
+//         let abi_decode_value0_57 := _3
+//         let abi_decode_value1_58 := _3
+//         let abi_decode_value2_59 := _3
+//         let abi_decode_value3 := _3
+//         let abi_decode_value4 := _3
+//         if slt(sub(_1019, _1021), 128)
+//         {
+//             revert(_3, _3)
+//         }
+//         {
+//             abi_decode_value0_57 := and(calldataload(_1021), 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF)
+//         }
+//         {
+//             abi_decode_value1_58 := calldataload(add(_1021, 32))
+//         }
+//         {
+//             let abi_decode_offset_62 := calldataload(add(_1021, 64))
+//             if gt(abi_decode_offset_62, 0xffffffffffffffff)
+//             {
+//                 revert(_3, _3)
+//             }
+//             abi_decode_value2_59, abi_decode_value3 := abi_decode_t_bytes_calldata_ptr(add(_1021, abi_decode_offset_62), _1019)
+//         }
+//         {
+//             abi_decode_value4 := cleanup_revert_t_enum$_Operation_$1949(calldataload(add(_1021, 96)))
+//         }
+//         sstore(_1, abi_decode_value0_57)
+//         sstore(_1, abi_decode_value1_58)
+//         sstore(_1, _3)
+//         sstore(_1, _3)
+//         sstore(_1, abi_decode_value4)
+//     }
+//     function abi_decode_t_bytes_calldata_ptr(offset_12, end_13) -> arrayPos_14, length_15
+//     {
+//         if iszero(slt(add(offset_12, 0x1f), end_13))
+//         {
+//             revert(0, 0)
+//         }
+//         let length_15_669 := calldataload(offset_12)
+//         length_15 := length_15_669
+//         if gt(length_15_669, 0xffffffffffffffff)
+//         {
+//             revert(0, 0)
+//         }
+//         arrayPos_14 := add(offset_12, 0x20)
+//         if gt(add(add(offset_12, length_15_669), 0x20), end_13)
```

Comment:
I we'd teach the optimizer to try to commute it could reuse arrayPos_14 here...

Reply:
Would probably require some sort of unification algorithm with backtracking or something like that.

Reply:
Yeah, sounds a little too complicated for now. In my devcon talk, I was talking about variables having linear dependencies on other variables, something like that could solve it.